source/s3_meta_request.c

```diff
@@ -233,6 +324,122 @@ static void s_s3_request_on_shutdown(void *user_data) {
     /*************** GIL RELEASE ***************/
 }
 
+/* TODO: probably should put the input stream to somewhere else, but to keep it simple and timely manner. */
+/*
+ * file-based python input stream for reporting the progress
+ */
+struct aws_input_py_stream_file_impl {
+    struct aws_input_stream *actual_stream;
+    struct s3_meta_request_binding *binding;
+};
+
+static int s_aws_input_stream_file_read(struct aws_input_stream *stream, struct aws_byte_buf *dest) {
+    struct aws_input_py_stream_file_impl *impl = stream->impl;
+    size_t pre_len = dest->len;
+
+    if (aws_input_stream_read(impl->actual_stream, dest)) {
+        return AWS_OP_ERR;
+    }
+
+    size_t actually_read = 0;
+    if (aws_sub_size_checked(dest->len, pre_len, &actually_read)) {
+        return AWS_OP_ERR;
+    }
+
+    bool report_progress;
+    struct s3_meta_request_binding *request_binding = impl->binding;
+    if (s_record_progress(request_binding, (uint64_t)actually_read, &report_progress)) {
+        return AWS_OP_ERR;
+    }
+
+    if (report_progress) {
+        /*************** GIL ACQUIRE ***************/
+        PyGILState_STATE state;
+        if (aws_py_gilstate_ensure(&state)) {
+            return AWS_OP_SUCCESS; /* Python has shut down. Nothing matters anymore, but don't crash */
+        }
+        PyObject *result =
+            PyObject_CallMethod(request_binding->self_py, "_on_progress", "(k)", request_binding->size_transferred);
+        if (!result) {
+            PyErr_WriteUnraisable(request_binding->self_py);
```

Comment:
probably want to raise an error and eventually return `AWS_OP_ERR`?
We can report errors in this particular callback

use this bad boy:
```
/**
 * Raise an AWS error corresponding to the current Python error.
 *
 * Prints the current Python error to stderr and clears the Python error indicator.
 * Finds an AWS error code corresponding to the current Python error (fallback is AWS_ERROR_UNKNOWN).
 * Invokes aws_raise_error() with that error code and always returns AWS_OP_ERR.
 *
 * The Python error indicator MUST be set and the GIL MUST be held when calling this function. */
int aws_py_raise_error(void);
```